torch/_dynamo/variables/user_defined.py

```diff
@@ -52,20 +58,54 @@ def as_python_constant(self):
     def python_type(self):
         return type(self.value)
 
+    def as_proxy(self):
+        return self.value
+
+    def __repr__(self):
+        return f"UserDefinedClassVariable({self.value})"
+
+    @staticmethod
+    @functools.lru_cache(None)
+    def _constant_fold_classes():
+        return {
+            torch.device,
+            torch.finfo,
+            torch.iinfo,
+            torch.Size,
+        }
+
+    @staticmethod
+    @functools.lru_cache(None)
+    def _in_graph_classes():
+        return set(tensortype_to_dtype.keys()) | {
+            torch.Tensor,
+            torch.cuda.Stream,
+            torch.cuda.Event,
+        }
+
+    def can_constant_fold_through(self):
+        return self.value in self._constant_fold_classes()
+
     def var_getattr(self, tx, name: str) -> "VariableTracker":
+        from .. import trace_rules
         from . import ConstantVariable
         from .builder import VariableBuilder
 
+        if name == "__name__":
+            return ConstantVariable.create(self.value.__name__)
+
         source = AttrSource(self.source, name) if self.source is not None else None
         try:
             obj = inspect.getattr_static(self.value, name)
         except AttributeError:
             obj = None
 
         if isinstance(obj, staticmethod):
-            return variables.UserFunctionVariable(
-                obj.__get__(self.value), source=source
-            )
+            func = obj.__get__(self.value)
+            if trace_rules.lookup(func) is not None:
+                return trace_rules.lookup(func).create_with_source(func, source=source)
+            else:
+                return variables.UserFunctionVariable(func, source=source)
```

Comment:
This will be further consolidated later, where I'll move the ```UserFunctionVariable``` case into the ```trace_rules.lookup``` check as well. Then ```lookup``` will be the first class API to decide if a function would be put in graph, skipped or inlined.